ntoskrnl/ke/i386/cpu.c

```diff
@@ -1359,62 +1438,397 @@ KiCoprocessorError(VOID)
+             * save area from the processor control
+             * region.
+             */
+            FsContext->FxSaveArea = &CurrentPrcb->NpxSaveArea;
+            ASSERT(FsContext->FxSaveArea);
+        }
+        else if (KeGetCurrentIrql() == APC_LEVEL)
+        {
+            /*
+             * This is APC level, allocate the necessary
+             * pool memory to hold floating point stuff.
+             */
+            FsContext->FxSaveArea = ExAllocatePoolWithTag(NonPagedPool, sizeof(FX_SAVE_AREA) + 8, TAG_FLOATING_POINT);
+            if (!FsContext->FxSaveArea)
+            {
+                /* Bail out if we failed */
+                return STATUS_INSUFFICIENT_RESOURCES;
+            }
+
+            /* We've allocated the floating point state in pool */
+            FsContext->PoolAllocated = TRUE;
+        }
+    }
+
+    /* Disable interrupts now */
+    _disable();
+
+    /* Get the Cr0 and mask out FPU flags */
+    Cr0 = __readcr0();
+    __writecr0(Cr0 & ~(CR0_MP | CR0_TS | CR0_EM));
+
+    /*
+     * If the NPX thread is not the same as the
+     * current calling thread then obtain the NPX
+     * registers for this context (if the NPX
+     * thread itself is not NULL that is).
+     */
+    if (FsContext->CurrentThread != CurrentPrcb->NpxThread)
+    {
+        if ((CurrentPrcb->NpxThread != NULL) &&
+            (CurrentPrcb->NpxThread->NpxState == NPX_STATE_LOADED))
+        {
+            /* Get the FX frame */
+            FxSaveAreaFrame = KiGetThreadNpxArea(CurrentPrcb->NpxThread);
+
+            /* Save the FPU state */
+            Ke386SaveFpuState(FxSaveAreaFrame);
+
+            /* NPX thread has lost its state */
+            CurrentPrcb->NpxThread->NpxState = NPX_STATE_NOT_LOADED;
+            FxSaveAreaFrame->NpxSavedCpu = 0;
+        }
+
+        /* The new NPX thread is the current thread */
+        CurrentPrcb->NpxThread = FsContext->CurrentThread;
+    }
+
+    /* Now get the FX frame from current thread */
+    FxSaveAreaFrame = KiGetThreadNpxArea(FsContext->CurrentThread);
+
+    /* Make sure it's not trash */
+    ASSERT(FxSaveAreaFrame);
+
+    /* Is NPX state for current thread loaded? */
+    if (FsContext->CurrentThread->NpxState == NPX_STATE_LOADED)
+    {
+        /*
+         * It is, have we gotten the acknowledgement
+         * that the floating point data is allocated
+         * in pool memory?
+         */
+        if (FsContext->PoolAllocated)
+        {
+           /* It is, save the FPU state */
+           Ke386SaveFpuState(FsContext->FxSaveArea);
+
+           /* And save the NPX CR0 state as well */
+           FsContext->FxSaveArea->Cr0NpxState = FxSaveAreaFrame->Cr0NpxState;
+           FsContext->FxSaveArea->NpxSavedCpu = 0;
+        }
+        else
+        {
+            /* It's not, save the control word then */
+            Ke386SaveCwState(FsControlWord);
+
+            /* Get MXCsr if XMMI is supported */
+            if (KeI386XMMIPresent && KeI386FxsrPresent)
```

Comment:
Can't we just obtain that info from CurrentPrcb->FeatureBits? I don't get why a separate variable is needed.
In future, other extended state support will be added here

Reply:
Wouldn't `CurrentPrcb` be the per-processor state?

Using the `KeI386...Present` *global* variables ensures consistency of usage with respect to all these other Ke386Save* and similar functions, where the initial *Present -variables state is the one of the boot processor:
https://git.reactos.org/?p=reactos.git;a=blob;f=ntoskrnl/ke/i386/kiinit.c;hb=75d33bf1481d328788e9898743421e907e7fc5fc#l469
and it is **expected** that all the other processors on an SMP system *have* the same characteristics as the boot processor (i.e. if the boot processor has FPU / XMM / whatever support, then **all** the other processors must have it).